File: c/easy/binary_gap.c
#include <stdio.h>

int binaryGap(int n) {

    // Remove trailing zeros so the first bit is 1
    n /= (n & -n);

    if (n == 1)
        return 0;

    int longest_distance = 0;
    int zeros_since_last_one = 0;

    while (n) {
        if (n & 1) {  // Found a '1' bit
            if (zeros_since_last_one > longest_distance)
                longest_distance = zeros_since_last_one;

            zeros_since_last_one = 0;
        } else {      // Counting zeros between 1s
            zeros_since_last_one++;
        }

        n >>= 1;
    }

    // Distance = zeros between + 1
    return longest_distance + 1;
}

File: rust/easy/binary_gap.rs
pub fn binary_gap(mut n: i32) -> i32 {

    // Remove trailing zeros so the first bit is 1
    n /= n & -n;

    if n == 1 {
        return 0;
    }

    let mut longest_distance = 0;
    let mut zeros_since_last_one = 0;

    while n != 0 {
        if (n & 1) == 1 {
            longest_distance =
                longest_distance.max(zeros_since_last_one);
            zeros_since_last_one = 0;
        } else {
            zeros_since_last_one += 1;
        }

        n >>= 1;
    }

    longest_distance + 1
}